### PR TITLE
Add option for leeway to account for clock skew times between the sig…

### PR DESCRIPTION
**Pull request by lats1**
…ning and verifying servers

### CODE REVIEW

File: src/Security/OpenIdConfigurationProvider.php [hunk 1]
```diff
@@ -56,6 +56,11 @@ class OpenIdConfigurationProvider extends AbstractProvider
      */
     private $responseResourceOwnerId = 'id';
 
+    /**
+     * @var int
+     */
+    private $leeway = 0;
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php]: As far as I can tell you never use this, only the local variable in `validateIdToken` is used?

File: src/Security/OpenIdConfigurationProvider.php [hunk 2]
```diff
@@ -192,10 +197,32 @@ public function getAuthorizationUrl(array $options = []): string
      *
      * @throws ItkOpenIdConnectException
      */
-    public function validateIdToken(string $idToken, string $nonce): object
+
+    /**
+     * Do any required verification of the id token and return an array of decoded claims
+     *
+     * @param string $idToken
+     *   Raw id token as string
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php]: Let's drop "as string", it's typehinted, so really don't know why I wrote it.

File: src/Security/OpenIdConfigurationProvider.php [hunk 3]
```diff
@@ -192,10 +197,32 @@ public function getAuthorizationUrl(array $options = []): string
...
+     *
+     * @param string $nonce
+     *   Nonce as string
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php]: Good catch on the missing docs parameter 👍 

(And again - Let's drop "as string")

File: src/Security/OpenIdConfigurationProvider.php [hunk 4]
```diff
@@ -192,10 +197,32 @@ public function getAuthorizationUrl(array $options = []): string
...
+     *
+     * @param int $leeway
+     *   Leeway set in seconds. Defaults to 0 and must be positive
+     *
+     * @return object
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php:197]: Why remove the description: "The JWT's payload as a PHP object"

File: src/Security/OpenIdConfigurationProvider.php [hunk 5]
```diff
@@ -192,10 +197,32 @@ public function getAuthorizationUrl(array $options = []): string
      *
      * @throws ItkOpenIdConnectException
      */
-    public function validateIdToken(string $idToken, string $nonce): object
+
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php]: You have duplicated the docblock for `validateIdToken`, please merge the two blocks

File: src/Security/OpenIdConfigurationProvider.php [hunk 6]
```diff
@@ -192,10 +197,32 @@ public function getAuthorizationUrl(array $options = []): string
...
+     *   Leeway set in seconds. Defaults to 0 and must be positive
+     *
+     * @return object
+     *
+     * @throws ItkOpenIdConnectException
+     */
+    public function validateIdToken(string $idToken, string $nonce, int $leeway = 0): object
     {
+        if ($leeway < 0) {
+            throw new \InvalidArgumentException('$leeway has to be a positive integer');
```
Comment (turegjorup) [inline src/Security/OpenIdConfigurationProvider.php]: Please don't use PHP's built-in exceptions directly. USe one of the exceptions in `src/Exception` or introduce a new one if needed.

The idea is that all exceptions thrown extend `ItkOpenIdConnectException`. That way a user of the lib can decide wheater to just catch `ItkOpenIdConnectException` or catch the specific exception.

....This properly deserves an entry in the readme.

Comment (turegjorup) [reply src/Security/OpenIdConfigurationProvider.php]: Also this is really easy to write at test for - just saying 😃

File: README.md [hunk 7]
```diff
@@ -142,6 +142,18 @@ try {
 }
 ```
 
+##### Leeway
+
+to account for clock skew times between the signing and verifying servers,
```
Comment (turegjorup) [inline README.md]: Maybe reference the `firebase/php-jwt` docs: "It is recommended that this leeway should not be bigger than a few minutes."

**Review by turegjorup — CHANGES_REQUESTED** [review]

File: src/Security/OpenIdConfigurationProvider.php (hunk 1 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php]: Fixed in 98bf5d0

File: src/Security/OpenIdConfigurationProvider.php (hunk 2 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php]: Fixed in 860c85a

File: src/Security/OpenIdConfigurationProvider.php (hunk 3 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php]: Fixed in 860c85a

File: src/Security/OpenIdConfigurationProvider.php (hunk 4 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php:197]: Fixed in 860c85a

File: src/Security/OpenIdConfigurationProvider.php (hunk 5 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php]: Fixed in 860c85a

File: src/Security/OpenIdConfigurationProvider.php (hunk 6 above)
Comment (lats1) [reply src/Security/OpenIdConfigurationProvider.php]: Fixed in 4804dbd, b49693c and in 81e595b.

File: README.md (hunk 7 above)
Comment (lats1) [reply README.md]: Fixed in 81e595b

**Review by turegjorup — APPROVED** [review]

**Pull request merged**